src/cpp/session/modules/rmarkdown/SessionExecuteChunkOperation.hpp

```diff
@@ -0,0 +1,308 @@
+
+#include "NotebookOutput.hpp"
+#include "NotebookExec.hpp"
+#include "SessionRmdNotebook.hpp"
+
+namespace rstudio {
+namespace session {
+namespace modules {
+namespace rmarkdown {
+namespace notebook {
+
+core::shell_utils::ShellCommand shellCommandForEngine(
+      const std::string& engine,
+      const core::FilePath& scriptPath)
+{
+   core::shell_utils::ShellCommand command(engine);
+   
+   if (engine == "Rscript")
+      command << "-f";
+   
+   command << scriptPath;
+   
+   return command;
+}
+
+class ExecuteChunkOperation : boost::noncopyable,
+                              public boost::enable_shared_from_this<ExecuteChunkOperation>
+{
+   typedef core::shell_utils::ShellCommand ShellCommand;
+   typedef core::system::ProcessCallbacks ProcessCallbacks;
+   typedef core::system::ProcessOperations ProcessOperations;
+   
+public:
+   static boost::shared_ptr<ExecuteChunkOperation> create(const std::string& docId,
+                                                          const std::string& chunkId,
+                                                          const ShellCommand& command)
+   {
+      boost::shared_ptr<ExecuteChunkOperation> pProcess =
+            boost::shared_ptr<ExecuteChunkOperation>(new ExecuteChunkOperation(
+                                                        docId,
+                                                        chunkId,
+                                                        command));
+      pProcess->registerProcess();
+      return pProcess;
+   }
+   
+private:
+   
+   ExecuteChunkOperation(const std::string& docId,
+                         const std::string& chunkId,
+                         const ShellCommand& command)
+      : isRunning_(false),
+        terminationRequested_(false),
+        docId_(docId),
+        chunkId_(chunkId),
+        command_(command)
+   {
+      using namespace core;
+      Error error = Success();
+      
+      // ensure staging directory
+      FilePath stagingPath = chunkOutputPath(
+               docId_,
+               chunkId_ + kStagingSuffix,
+               ContextExact);
+      
+      error = stagingPath.removeIfExists();
+      if (error)
+         LOG_ERROR(error);
+      
+      error = stagingPath.ensureDirectory();
+      if (error)
+         LOG_ERROR(error);
+      
+      // ensure regular directory
+      FilePath outputPath = chunkOutputPath(
+               docId_,
+               chunkId_,
+               ContextExact);
+      
+      error = outputPath.removeIfExists();
+      if (error)
+         LOG_ERROR(error);
+      
+      error = outputPath.ensureDirectory();
+      if (error)
+         LOG_ERROR(error);
+      
+      // clean old chunk output
+      error = cleanChunkOutput(docId_, chunkId_, true);
+      if (error)
+         LOG_ERROR(error);
+   }
+   
+public:
+   
+   ProcessCallbacks processCallbacks()
+   {
+      ProcessCallbacks callbacks;
+      
+      callbacks.onStarted = boost::bind(
```

Comment:
FYI since you don't use `ProcessOperations` you can exclude it from all the callback signatures and then just don't pass `_1` to bind.